src/lib.rs

```diff
@@ -60,98 +41,251 @@ fn set_stack_limit(l: usize) {
+            // 32-bit pushes an argument so we need a bit of an offset to get it 16-byte
+            // aligned when the call is made.
+            let offset = if cfg!(target_pointer_width = "32") {
+                12
+            } else {
+                0
+            };
+
+            unsafe {
+                __stacker_switch_stacks(&mut f as *mut &mut FnMut() as *mut u8,
+                                        doit as usize as *const _,
+                                        stack_low + stack_size - offset);
+            }
+
+            // Dropping `switch` frees the memory mapping and restores the old stack limit
+        }
     }
 }
 
+extern fn doit(f: &mut &mut FnMut()) {
+    f();
+}
+
 cfg_if! {
     if #[cfg(windows)] {
-        // See this for where all this logic is coming from.
-        //
-        // https://github.com/adobe/webkit/blob/0441266/Source/WTF/wtf
-        //                   /StackBounds.cpp
-        unsafe fn guess_os_stack_limit() -> usize {
-            #[cfg(target_pointer_width = "32")]
-            extern {
-                #[link_name = "__stacker_get_tib_32"]
-                fn get_tib_address() -> *const usize;
+        extern {
+            fn __stacker_get_current_fiber() -> winapi::PVOID;
+        }
+
+        #[no_mangle]
+        pub unsafe extern fn __stacker_switch_stacks_callback(f: &mut &mut FnMut()) {
+            f();
+        }
+
+        struct FiberInfo<'a> {
+            callback: &'a mut FnMut(),
+            result: Option<std::thread::Result<()>>,
+            parent_fiber: winapi::LPVOID,
+        }
+
+        unsafe extern "system" fn fiber_proc(info: winapi::LPVOID) {
+            let info = &mut *(info as *mut FiberInfo);
+            info.result = Some(std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                (info.callback)();
+            })));
+            kernel32::SwitchToFiber(info.parent_fiber);
+            return;
+        }
+
+        fn _grow(stack_size: usize, callback: &mut FnMut()) {
+            unsafe {
+                let was_fiber = kernel32::IsThreadAFiber() == winapi::TRUE;
+                
+                let mut info = FiberInfo {
+                    callback,
+                    result: None,
+                    parent_fiber: if was_fiber {
+                        __stacker_get_current_fiber()
+                    } else {
+                        kernel32::ConvertThreadToFiber(0i32 as _)
+                    },
+                };
+                if info.parent_fiber == 0i32 as _ {
+                    panic!("unable to convert thread to fiber");
+                }
+                let fiber = kernel32::CreateFiber(stack_size as _, Some(fiber_proc), &mut info as *mut FiberInfo as *mut _);
+                if fiber == 0i32 as _ {
+                    panic!("unable to allocate fiber");
+                }
+                kernel32::SwitchToFiber(fiber);
+                kernel32::DeleteFiber(fiber);
```

Comment:
This seems... suspicious. I don't know what fibers are, but it may be worth a comment saying why we're deleting the thing we switched to, as we're seemingly switched to it.

Reply:
So here we switch to our newly created coroutine. When it returns we just delete it.